### PR TITLE
Add proposed custom code guidelines

### PR DESCRIPTION
**Pull request by esoergel**
I wrote this a while back and got some positive feedback on it.  I'd like to add this to the repo so we can point to it next time it becomes relevant.

Let me know your thoughts - are these stipulations fair to impose as requirements on future custom work?

### CODE REVIEW

**Review by dannyroberts — COMMENTED** [review]
Looks good to me. Added Clayton as a reviewer with his GTD hat on, in case there's a consideration the rest of us might miss in framing this as a open source standard vs a Dimagi team standard or something like that.

**Review by snopoke — COMMENTED** [review]
Looks good. Can we add a note about not relying on implementation details of other parts of HQ and only using 'public' APIs.

**Comment by proteusvacuum** [comment]
Looks good!

**Comment by ctsims** [comment]
I'd like to do a pass here which frames this a bit less explicitly around Dimagi and orients towards a credible Open Source story. Will try to wrap that up now, but if I don't please reach back out before approving.

**Comment by ctsims** [comment]
@esoergel Did a pass integrating Simon's suggestion, but mostly reframing a little bit away from being a "Dimagi" guideline into being an open source guideline

File: custom/CONTRIBUTING.rst [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+a clear low-risk path to maintenance. 
+
+Support
+-------
+When custom code is developed by a contributor who is not a core maintainer, 
+they should generally be connected with a project maintainer for backstopping 
+before beginning development.
+
+This person can answer questions about how to meet the above requirements, and
+how much effort is practical to require The contributor should check in regularly
```
Comment (dannyroberts) [inline custom/CONTRIBUTING.rst:123]:
```suggestion
how much effort is practical to require. The contributor should check in regularly
```

**Review by dannyroberts — APPROVED** [review]

File: custom/CONTRIBUTING.rst [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+The code should be explicit about how and where it is run. It should be
+registered in ``DOMAIN_MODULE_MAP``, so it’s clear which domains are expected to
+use the code, and it should be restricted to run only on those domains, and on a
+specific server environment (or multiple, if necessary).
+
+Dependencies
+------------
+Custom code should rely on as few dependencies as possible which will increase
+the burden of maintenance long term. This means that the code should rely only
+on the common and public interfaces to the rest of the core code, and stricly
```
Comment (esoergel) [inline custom/CONTRIBUTING.rst:108]:
```suggestion
on the common and public interfaces to the rest of the core code, and strictly
```

File: custom/CONTRIBUTING.rst [hunk 3]
```diff
@@ -78,16 +93,36 @@ registered in ``DOMAIN_MODULE_MAP``, so it’s clear which domains are expected
...
 Support
 -------
-When custom code is developed by an external contractor, they should have access
-to a Dimagi developer for backstopping. This person can answer questions about
-how to meet the above requirements, or decide that it’s okay to waive one or
-more in some particular circumstance. The contractor should check in regularly
-during the development process with the Dimagi developer to ensure the approach
+When custom code is developed by a contributor who is not a core maintainer, 
+they should generally be connected with a project maintainer for backstopping 
+before beginning development.
```
Comment (esoergel) [inline custom/CONTRIBUTING.rst:120]: nice framing

File: custom/CONTRIBUTING.rst [hunk 4]
```diff
@@ -17,7 +17,15 @@ sure to put in the effort to ensure it can be maintained, understood, and
...
-
+It is important to note upfront that although CommCare's code is an Open Source
+community good, the willingness of the code maintainers to accept custom code
+contributions should be considered a privilege and not a right. Since custom 
+code creates a permanent drag on the organizations and individuals who maintain
+the software, the project maintainers are free to use their judgement in the
+decision of whether that long term cost is warranted, just as they are in the
+determination of whether new features are appropriate. These guidelines should
+be interpreted as neccessary, but not sufficient, for a custom contribution's
+inclusion.
```
Comment (esoergel) [inline custom/CONTRIBUTING.rst:28]: :+1:

**Comment by esoergel** [comment]
Great additions, @ctsims , thanks!

**Review by millerdev — APPROVED** [review]
Excellent!

**Pull request merged**